File: src/main/java/aoc2021/Day2.kt
package aoc2021


fun main() = Day2().run()
class Day2 : Day(2) {
    override fun part1(input: List<String>) : Any {
        var x = 0
        var y = 0
        for(line : String in input){
            val groups = line.split("\\s")
            val n = groups[1].toInt()
            when(groups[0]){
                "forward" -> x += n
                "up" -> y -= n
                "down" -> y += n
                else -> throw IllegalArgumentException()
            }
        }
        return x * y
    }

    override fun part2(input: List<String>) : Any{
        var x = 0
        var y = 0
        var aim = 0
        for(line in input){
            val groups = line.split("\\s")
            val n = groups[1].toInt()
            when(groups[0]){
                "forward" -> {x += n
                    y += aim * n}
                "up" -> aim -= n
                "down" -> aim += n
                else -> throw IllegalArgumentException()
            }
        }
        return x * y
    }
}

File: src/main/java/aoc2022/Day2.scala
package aoc2022


import scala.language.postfixOps

object Day2 extends Day(2) {
  def main(args: Array[String]): Unit = super.run()

  val guide: List[(Int, Int)] = input.map(line => (line(0) - 'A' + 1, line(2) - 'X' + 1))

  //1 = Rock, 2 = paper, 3 = scissors
  val moves: Seq[Int] = Seq(1, 2, 3, 1, 2)

  override def part1(): Any = {
    guide.map((oppMove, playerMove) =>
      playerMove + 3 * {
        if (moves(moves.indexOf(oppMove) + 1) == playerMove) 2
        else if (playerMove == oppMove) 1
        else 0
      }).sum
  }

  override def part2(): Any = {
    guide.map((oppMove, result) => (result match {
      case 1 => moves(moves.indexOf(oppMove) + 2)
      case 2 => oppMove
      case 3 => moves(moves.indexOf(oppMove) + 1)
    }) + 3 * (result - 1)).sum
  }
}
